src/vmm/src/persist.rs

```diff
@@ -247,6 +251,38 @@ fn snapshot_memory_to_file(
     }
 }
 
+/// Validates that snapshot CPU vendor matches the host CPU vendor.
+#[cfg(target_arch = "x86_64")]
+pub fn validate_cpu_vendor(
+    microvm_state: &MicrovmState,
+) -> std::result::Result<(), LoadSnapshotError> {
+    let host_vendor_id = get_vendor_id().map_err(|_| {
+        LoadSnapshotError::CpuVendorMismatch("Failed to read vendor from CPUID.".to_owned())
+    })?;
+
+    // Search for vendor id entry in state CPUID.
+    for entry in microvm_state.vcpu_states[0].cpuid.as_slice().iter() {
+        if entry.function == 0 && entry.index == 0 {
+            let cpu_vendor_id: [u8; 12] =
+                unsafe { std::mem::transmute([entry.ebx, entry.edx, entry.ecx]) };
+
+            if host_vendor_id != cpu_vendor_id {
+                let error_string = format!(
+                    "Host CPU vendor id: {:?}, Snapshot CPU vendor id: {:?}",
+                    &host_vendor_id, &cpu_vendor_id
+                );
+                error!("{}", error_string);
+                return Err(LoadSnapshotError::CpuVendorMismatch(error_string));
+            } else {
+                info!("Snapshot CPU vendor id: {:?}", &cpu_vendor_id);
```

Comment:
nit: function is in the hotpath so move the logging outside the loop, it's enough to print the vendor id once instead of repeating it for each vcpu.

Reply:
We only check the first vCPU. The cpuid entry we are looking should be the first entry and the looping stops once it fails/passes the check.

Reply:
right, I misread the code